packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,330 @@
+		const result: VisualValueNode = {
+			value: tree.value,
+			nodeKind: VisualNodeKind.ValueNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+		return result;
+	} else {
+		const children: Record<string, VisualChildNode> = {};
+
+		for (const [key, value] of Object.entries(tree.fields)) {
+			const child = toVisualTree(value);
+			children[key] = child;
+		}
+
+		return {
+			children,
+			nodeKind: VisualNodeKind.TreeNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+	}
+}
+
+/**
+ * Returns the allowed fields & types for the object fields.
+ */
+function getObjectAllowedTypes(schema: ObjectNodeStoredSchema): string {
+	let result = "";
+
+	for (const [fieldKey, treeFieldStoredSchema] of schema.objectNodeFields) {
+		// Set of allowed tree types {@link TreeTypeSet}.
+		const fieldTypes = treeFieldStoredSchema.types;
+
+		let fieldAllowedType = fieldKey === "" ? "" : `${fieldKey} : `;
+
+		// If not specified, types are unconstrained.
+		if (fieldTypes === undefined) {
+			fieldAllowedType += "any";
+		} else {
+			for (const type of fieldTypes) {
+				fieldAllowedType += `${type} | `;
+			}
+		}
+
+		// Slice the trailing ` | ` from the `fieldAllowedType`.
+		fieldAllowedType = `${fieldAllowedType.slice(0, -3)}, `;
+
+		result += fieldAllowedType;
+	}
+
+	// Slice the trailing `, ` from the `result`.
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+/**
+ * Returns the allowed fields & types for the map fields.
+ */
+function getMapAllowedTypes(
+	fields: FieldMapObject<JsonableTree> | undefined,
+	schema: MapNodeStoredSchema,
+): string {
+	let fieldAllowedTypes = "";
+
+	const mapFieldAllowedTypes = schema.mapFields.types;
+
+	if (mapFieldAllowedTypes === undefined) {
+		fieldAllowedTypes = "any";
+	} else {
+		for (const type of mapFieldAllowedTypes) {
+			fieldAllowedTypes += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `fieldAllowedType`.
+	fieldAllowedTypes = fieldAllowedTypes.slice(0, -3);
+
+	if (fields === undefined) {
+		throw new TypeError("Fields should not be undefined.");
+	}
+
+	let result = "";
+	for (const [fieldKey] of Object.entries(fields)) {
+		result += `${fieldKey} : ${fieldAllowedTypes}, `;
+	}
+
+	// Slice the trailing `, ` from the `result`.
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+/**
+ * Returns the allowed fields & types for the leaf fields.
+ */
+function getLeafAllowedTypes(schema: TreeFieldStoredSchema): string {
+	let result = "";
+	const leafTypes = schema.types;
+
+	if (leafTypes === undefined) {
+		result = "any";
+	} else {
+		for (const type of leafTypes) {
+			result += `${type} | `;
+		}
+	}
+
+	// Slice the trailing ` | ` from the `result`.
+	result = `${result.slice(0, -3)}`;
+
+	return result;
+}
+
+/**
+ * Returns the schema & leaf value of the node with type {@link LeafNodeStoredSchema}.
+ */
+function visualizeLeafNodeStoredSchema(
+	tree: JsonableTree,
+	allowedTypes: string,
```

Comment:
I'm not sure it makes sense to have `allowedTypes` on a leaf node. `allowedTypes` describes what kinds of data are allowed _under_ the node (i.e. as children). Leaves can't have children, so I don't think they need to specify allowed types. Just specifying the schema name and type for leaves is probably sufficient.

Reply:
The parent node will already say what the allowed types are for the field this leaf lives under, so that information is already captured (appropriately) 1 level higher.

Reply:
This seems to be the area where people's opinions diverge (and I like it 😄 ). 

As far as I remember, @kashms & @alexvy86 were in favor of displaying which data types are allowed even at the leaf node. So referring our current test cases (which will be fixed before this PR gets merged), leaf node `baz` will have `[builder.number, builder.string, builder.boolean]` as its allowed types. 

Strictly speaking in the Shared Tree POV, your suggestion is correct. Since there data cannot live under the leaf node, `allowedTypes` should only be available until one level higher than the leaf node. And Shared Tree API does not allow schema of type `LeafNodeStoredSchema` to access these data which corresponds to your point. 

However, my understanding after several discussions during parking lot / meetings was that _despite_ such API limitations, we would still want to add this data at the leaf node.

Below are some pros and cons that I have realized working on this feature: 

Pros:
- If a developer using devtools is interested in the allowed data type of the leaf node he or she doesn't have to click additional layer to check the allowed data type. 

Cons:
- In terms of maintaining the readability / logic of the `SharedTreeVisualizer`, adding allowed types for the leaf node adds several burdens.
    - This is not supported in the SharedTree, thus we need to first check if the child node is a type of `LeafNodeStoredSchema` and then apply different logics for `array`, `object` and `map` node (Check line `195`, `218` and `276` from the `SharedTreeVisualizer.ts`.
    - Due to the API constraints, the pre-constructed string is then passed into the main recursive function `visualizeSharedTreeNodeBySchema` as an optional string (which would be confusing in future occasions). 
    - Although this is unlikely, if the SharedTree is a single leaf node, we should also apply the same logic listed above in the `DefaultVisualizers.ts`'s `VisualizeSharedTree` before calling the `visualizeSharedTreeNodeBySchema()`

Reply:
Sorry, I didn't offer sufficient context / clarity here. I'm not suggesting we don't display something against leaves - that information is useful. I'm suggesting that "allowedTypes" is not the correct thing to display. "allowed types" describes something at a contextually different level. I think about it this way:
1. "schema" describes the schema associated with the data at the same level.
2. "allowed types" describes the types of things allowed _under_ the level of the node.

So, I'm suggesting we shouldn't pass down the "allowedTypes" from the parent to leaf children. The allowed types are a concept specific to the parent node. Instead, for leaves, I think we should display the _schema_ associated with that leaf node. Presumably you can get this from the leaf node itself.

At a high level, the schema information we display at a given level in the tree should describe the schema of that node, not what kinds of things are allowed by the parent. If a developer wants to view that information, they can view the schema of the parent node.

Reply:
+1 to Josh's clarification. I guess I just don't remember off the top of my head, where exactly in that "parent level" (where we have access to the field) are we displaying the allowed types for the field.

Assuming a tree where the root has 3 fields foo (string/number), bar (array), baz (map of string to boolean), I think something like this makes sense:

```
SharedTree
  data:
    tree:
      foo (string/number):
        value (number): 32
      bar (array of string):
        0:
          value (string): "hello"
        1:
          value (string): "world"
      baz (map of boolean):
        "key1":
          value (boolean): true
        "key2":
          value (boolean): false
```